// redux-tools/utils.js
/* global navigator */
var backupStorageData = {};
export var backupStorage = {
  clear: function clear() {
    Object.keys(backupStorageData).forEach(function (key) {
      delete backupStorageData[key];
    });
  },
  getItem: function getItem(key) {
    return backupStorageData[key] || null;
  },
  removeItem: function removeItem(key) {
    delete backupStorageData[key];
  },
  setItem: function setItem(key, value) {
    backupStorageData[key] = value;
  }
};
export var platforms = {
  BROWSER: 'BROWSER',
  NODE: 'NODE',
  REACT_NATIVE: 'REACT_NATIVE'
};
export var getPlatform = function getPlatform() {
  if (typeof document !== 'undefined') {
    return platforms.BROWSER;
  } else if (typeof navigator !== 'undefined' && navigator.product === 'ReactNative') {
    return platforms.REACT_NATIVE;
  }

  return platforms.NODE;
};
export var snakeToCamelCase = function snakeToCamelCase(str) {
  return str.toLowerCase().replace(/_[0-9A-Z_a-z]/g, function (match) {
    return match[1].toUpperCase();
  });
};
export default {
  backupStorage: backupStorage,
  getPlatform: getPlatform,
  platforms: platforms,
  snakeToCamelCase: snakeToCamelCase
};
